File: WaterSort/tools/level_water_solver.py
"""BFS 最少步数求解（与 LevelWaterSolver.cs 规则一致）。"""

from __future__ import annotations

import math
from collections import deque
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

MAX_CAPACITY = 4
POCKET_COUNT = 4
MAX_DEPTH = 250
MAX_VISITED = 120000


@dataclass
class SolveResult:
    is_solvable: bool = False
    min_steps: int = 0
    search_limit_steps: int = 0
    message: str = ""


@dataclass
class BottleSim:
    is_lock_bottle: bool = False
    lock_color: int = 0
    lock_remaining: int = 0
    wh_nums: int = 0
    packed: bool = False
    stack: List[int] = field(default_factory=list)

    @property
    def in_play(self) -> bool:
        return not self.packed

    @property
    def is_locked(self) -> bool:
        return self.is_lock_bottle and self.lock_remaining > 0


class SolverState:
    def __init__(self) -> None:
        self.bottles: List[BottleSim] = []
        self.collected = 0
        self.pocket_colors = [0] * POCKET_COUNT
        self.pocket_locked = [False, False, True, True]
        self.pocket_refill: List[int] = []

    def clone(self) -> "SolverState":
        s = SolverState()
        s.collected = self.collected
        s.pocket_colors = list(self.pocket_colors)
        s.pocket_locked = list(self.pocket_locked)
        s.pocket_refill = list(self.pocket_refill)
        for b in self.bottles:
            s.bottles.append(BottleSim(
                is_lock_bottle=b.is_lock_bottle,
                lock_color=b.lock_color,
                lock_remaining=b.lock_remaining,
                wh_nums=b.wh_nums,
                packed=b.packed,
                stack=list(b.stack),
            ))
        return s

    def serialize_key(self) -> str:
        parts = [f"C{self.collected};"]
        for p in range(POCKET_COUNT):
            parts.append(f"{self.pocket_colors[p]}{'L' if self.pocket_locked[p] else 'U'},")
        parts.append("|")
        parts.extend(f"{r}," for r in self.pocket_refill)
        parts.append("|")
        for b in self.bottles:
            parts.append("P" if b.packed else "A")
            parts.append(f"{b.lock_remaining}:")
            parts.extend(f"{c}," for c in b.stack)
            parts.append(";")
        return "".join(parts)

    def is_hidden_layer(self, bi: int, li: int) -> bool:
        b = self.bottles[bi]
        return b.wh_nums > 0 and len(b.stack) > 1 and li < b.wh_nums

    def is_collect(self, i: int) -> bool:
        b = self.bottles[i]
        if not b.in_play or len(b.stack) != MAX_CAPACITY or b.wh_nums > 0:
            return False
        c = b.stack[0]
        return all(x == c for x in b.stack)

    def can_pour_from(self, i: int) -> bool:
        b = self.bottles[i]
        return b.in_play and not b.is_locked and b.stack and not self.is_collect(i)

    def can_pour_to(self, i: int) -> bool:
        b = self.bottles[i]
        return b.in_play and not b.is_locked and len(b.stack) < MAX_CAPACITY

    def can_pour(self, fr: int, to: int) -> bool:
        if not self.can_pour_from(fr) or not self.can_pour_to(to):
            return False
        if not self.bottles[to].stack:
            return True
        return self.bottles[fr].stack[-1] == self.bottles[to].stack[-1]

    def top_pour_run(self, fr: int) -> int:
        stack = self.bottles[fr].stack
        if not stack:
            return 0
        color = stack[-1]
        run = 0
        for i in range(len(stack) - 1, -1, -1):
            if self.is_hidden_layer(fr, i):
                break
            if stack[i] != color:
                break
            run += 1
        return run

    def apply_pour(self, fr: int, to: int, move: int) -> None:
        for _ in range(move):
            color = self.bottles[fr].stack.pop()
            self.bottles[to].stack.append(color)

    def has_open_pocket_for(self, color: int) -> bool:
        for p in range(POCKET_COUNT):
            if not self.pocket_locked[p] and self.pocket_colors[p] == color:
                return True
        return False

    def unlock_next_pocket(self) -> bool:
        for p in range(POCKET_COUNT):
            if not self.pocket_locked[p]:
                continue
            self.pocket_locked[p] = False
            self.pocket_colors[p] = self.pocket_refill[0] if self.pocket_refill else 0
            if self.pocket_refill:
                self.pocket_refill.pop(0)
            return True
        return False

    def try_unlock_pocket_for_pending_collect(self) -> bool:
        for i, b in enumerate(self.bottles):
            if not self.is_collect(i) or b.is_locked:
                continue
            color = b.stack[-1]
            if self.has_open_pocket_for(color):
                continue
            return self.unlock_next_pocket()
        return False

    def apply_lock_unlock_on_pack(self, packed_color: int) -> None:
        for b in self.bottles:
            if not b.is_lock_bottle or b.lock_remaining <= 0:
                continue
            if b.lock_color != 0 and b.lock_color != packed_color:
                continue
            b.lock_remaining -= 1

    def pack_bottle(self, bi: int, pi: int) -> None:
        color = self.bottles[bi].stack[-1]
        self.bottles[bi].packed = True
        self.bottles[bi].stack.clear()
        self.collected += 1
        self.apply_lock_unlock_on_pack(color)
        self.pocket_colors[pi] = self.pocket_refill[0] if self.pocket_refill else 0
        if self.pocket_refill:
            self.pocket_refill.pop(0)

    def try_pack_one(self) -> bool:
        for p in range(POCKET_COUNT):
            if self.pocket_locked[p]:
                continue
            need = self.pocket_colors[p]
            if need <= 0:
                continue
            for i, b in enumerate(self.bottles):
                if not self.is_collect(i) or b.is_locked:
                    continue
                if b.stack[-1] != need:
                    continue
                self.pack_bottle(i, p)
                return True
        return False

    def apply_auto_pack_and_unlock(self) -> None:
        for _ in range(64):
            if self.try_pack_one():
                continue
            if not self.try_unlock_pocket_for_pending_collect():
                break


def _shuffle(lst: List[int], seed: int) -> None:
    rng = __import__("random").Random(seed)
    for i in range(len(lst) - 1, 0, -1):
        j = rng.randint(0, i)
        lst[i], lst[j] = lst[j], lst[i]


def _build_initial(cups: List[dict], level_index: int) -> Tuple[Optional[SolverState], int, Optional[str]]:
    state = SolverState()
    pocket_build: List[int] = []
    color_count: Dict[int, int] = {}
    total_layers = 0

    for cup in cups:
        if cup.get("isNull") or cup.get("isVideo"):
            continue
        sim = BottleSim()
        if cup.get("isLock"):
            sim.is_lock_bottle = True
            sim.lock_color = cup.get("lockColor", 0)
            sim.lock_remaining = cup.get("lockNums", 0) or 1
        sim.wh_nums = cup.get("whNums", 0)
        colors = cup.get("colors") or []
        sim.stack = list(colors)
        total_layers += len(colors)
        for c in colors:
            color_count[c] = color_count.get(c, 0) + 1
            if color_count[c] % 4 == 0:
                pocket_build.append(c)
        state.bottles.append(sim)

    if not state.bottles:
        return None, 0, "没有参与求解的瓶子"
    if total_layers % 4:
        return None, 0, "参与瓶水层总数不是 4 的倍数"

    need_collect = total_layers // 4
    if level_index != 1:
        _shuffle(pocket_build, level_index * 7919 + 17)

    state.pocket_refill = pocket_build
    for p in range(POCKET_COUNT):
        if state.pocket_locked[p]:
            state.pocket_colors[p] = 0
        else:
            state.pocket_colors[p] = state.pocket_refill[0] if state.pocket_refill else 0
            if state.pocket_refill:
                state.pocket_refill.pop(0)

    state.collected = 0
    state.apply_auto_pack_and_unlock()
    return state, need_collect, None


def try_find_min_steps(
    cups: List[dict],
    level_index: int = 1,
    max_depth: int = MAX_DEPTH,
    max_visited: int = MAX_VISITED,
) -> SolveResult:
    result = SolveResult()
    if not cups:
        result.message = "无瓶子数据"
        return result

    initial, need_collect, err = _build_initial(cups, level_index)
    if err:
        result.message = err
        return result
    assert initial is not None

    if need_collect == 0:
        result.is_solvable = True
        return result

    visited = {initial.serialize_key()}
    queue: deque = deque([(initial, 0)])

    while queue:
        state, depth = queue.popleft()
        if state.collected >= need_collect:
            result.is_solvable = True
            result.min_steps = depth
            return result
        if depth >= max_depth:
            continue
        if len(visited) >= max_visited:
            break

        n = len(state.bottles)
        for fr in range(n):
            if not state.can_pour_from(fr):
                continue
            run = state.top_pour_run(fr)
            for to in range(n):
                if fr == to or not state.can_pour_to(to) or not state.can_pour(fr, to):
                    continue
                move = min(run, MAX_CAPACITY - len(state.bottles[to].stack))
                if move <= 0:
                    continue
                nxt = state.clone()
                nxt.apply_pour(fr, to, move)
                nxt.apply_auto_pack_and_unlock()
                key = nxt.serialize_key()
                if key in visited:
                    continue
                visited.add(key)
                queue.append((nxt, depth + 1))

    result.is_solvable = False
    result.search_limit_steps = max_depth
    result.message = (
        f"未在 {max_depth} 步 / {max_visited} 状态内找到解"
        if len(visited) >= max_visited
        else f"未在 {max_depth} 步内找到解"
    )
    return result

File: WaterSort/Assets/Editor/LevelEditor/LevelWaterSolver.cs
using System.Collections.Generic;
using System.Text;
using AsGame.Core;
using AsGame.Data;
using UnityEngine;
using XrCode;

namespace AsGame.Editor.LevelEditor
{
    public struct LevelSolveResult
    {
        public bool IsSolvable;
        public int MinSteps;
        public int SearchLimitSteps;
        public string Message;
    }

    /// <summary>
    /// 按正常游戏流程 BFS：倒水步数 + 自动装袋/锁瓶解锁（装袋不计步）。
    /// 参与：普通瓶、锁瓶、空瓶；不参与：空槽、广告瓶。
    /// </summary>
    public static class LevelWaterSolver
    {
        const int MaxCapacity = GameConstants.WaterMaxCount;
        const int PocketCount = 4;
        const int MaxDepth = 250;
        const int MaxVisited = 120000;

        public static LevelSolveResult TryFindMinSteps(
            IList<CupData> cups,
            int levelIndex = 1,
            int maxStepsLimit = 0,
            int maxVisitedLimit = 0)
        {
            var result = new LevelSolveResult();
            if (maxStepsLimit <= 0)
                maxStepsLimit = MaxDepth;
            var visitedLimit = maxVisitedLimit > 0
                ? maxVisitedLimit
                : maxStepsLimit <= LevelWaterCheckPolicy.MaxAllowedMinSteps
                    ? 30000
                    : MaxVisited;
            if (cups == null || cups.Count == 0)
            {
                result.Message = "无瓶子数据";
                return result;
            }

            if (!TryBuildInitialState(cups, levelIndex, out var initial, out var needCollect, out var error))
            {
                result.Message = error;
                return result;
            }

            if (needCollect == 0)
            {
                result.IsSolvable = true;
                result.MinSteps = 0;
                return result;
            }

            var startKey = initial.SerializeKey();
            var queue = new Queue<(SolverState state, int depth)>();
            var visited = new HashSet<string> { startKey };
            queue.Enqueue((initial.Clone(), 0));

            while (queue.Count > 0)
            {
                var (state, depth) = queue.Dequeue();
                if (state.Collected >= needCollect)
                {
                    result.IsSolvable = true;
                    result.MinSteps = depth;
                    return result;
                }

                if (depth >= maxStepsLimit)
                    continue;

                if (visited.Count >= visitedLimit)
                    break;

                for (var from = 0; from < state.Bottles.Count; from++)
                {
                    if (!state.CanPourFrom(from)) continue;
                    var pourRun = state.TopPourRun(from);
                    for (var to = 0; to < state.Bottles.Count; to++)
                    {
                        if (from == to || !state.CanPourTo(to)) continue;
                        if (!state.CanPour(from, to)) continue;

                        var move = Mathf.Min(pourRun, MaxCapacity - state.Bottles[to].Stack.Count);
                        if (move <= 0) continue;

                        var next = state.Clone();
                        next.ApplyPour(from, to, move);
                        next.ApplyAutoPackAndUnlock();

                        var key = next.SerializeKey();
                        if (!visited.Add(key)) continue;
                        queue.Enqueue((next, depth + 1));
                    }
                }
            }

            result.IsSolvable = false;
            result.SearchLimitSteps = maxStepsLimit;
            result.Message = visited.Count >= visitedLimit
                ? $"未在 {maxStepsLimit} 步 / {visitedLimit} 状态内找到解"
                : $"未在 {maxStepsLimit} 步内找到解（单关上限 {LevelWaterCheckPolicy.MaxAllowedMinSteps}）";
            return result;
        }

        static bool TryBuildInitialState(
            IList<CupData> cups,
            int levelIndex,
            out SolverState state,
            out int needCollect,
            out string error)
        {
            state = new SolverState();
            needCollect = 0;
            error = null;

            var totalLayers = 0;

            for (var i = 0; i < cups.Count; i++)
            {
                var cup = cups[i];
                if (cup == null || cup.isNull != 0 || cup.isVideo != 0)
                    continue;

                var sim = new BottleSim { SlotIndex = i };
                if (cup.isLock != 0)
                {
                    sim.IsLockBottle = true;
                    sim.LockColor = cup.lockColor;
                    sim.LockRemaining = cup.lockNums > 0 ? cup.lockNums : 1;
                }

                sim.WhNums = cup.whNums;
                sim.WhMask = CupWhLayerUtility.GetMask(cup);
                if (cup.colors != null)
                {
                    sim.Stack.AddRange(cup.colors);
                    totalLayers += cup.colors.Count;
                }

                state.Bottles.Add(sim);
            }

            if (state.Bottles.Count == 0)
            {
                error = "没有参与求解的瓶子";
                return false;
            }

            if (totalLayers % 4 != 0)
            {
                error = "参与瓶水层总数不是 4 的倍数";
                return false;
            }

            needCollect = totalLayers / 4;
            var pocketBuild = new List<int>();
            LevelPocketColorPlanner.BuildPocketColors(
                cups, levelIndex, pocketBuild, useUnityRandom: false);

            state.PocketRefill = pocketBuild;
            state.PocketLocked = new[] { false, false, true, true };
            state.PocketColors = new int[PocketCount];
            AssignInitialPocketColors(state);
            state.Collected = 0;
            state.ApplyAutoPackAndUnlock();
            return true;
        }

        static void AssignInitialPocketColors(SolverState state)
        {
            for (var p = 0; p < PocketCount; p++)
            {
                if (state.PocketLocked[p])
                {
                    state.PocketColors[p] = 0;
                    continue;
                }

                state.PocketColors[p] = state.PocketRefill.Count > 0 ? state.PocketRefill[0] : 0;
                if (state.PocketRefill.Count > 0)
                    state.PocketRefill.RemoveAt(0);
            }
        }

        sealed class BottleSim
        {
            public int SlotIndex;
            public bool Packed;
            public bool IsLockBottle;
            public int LockColor;
            public int LockRemaining;
            public int WhNums;
            public int WhMask;
            public List<int> Stack = new();

            public bool InPlay => !Packed;
            public bool IsLocked => IsLockBottle && LockRemaining > 0;

            public BottleSim CloneShallow()
            {
                return new BottleSim
                {
                    SlotIndex = SlotIndex,
                    Packed = Packed,
                    IsLockBottle = IsLockBottle,
                    LockColor = LockColor,
                    LockRemaining = LockRemaining,
                    WhNums = WhNums,
                    WhMask = WhMask,
                    Stack = new List<int>(Stack)
                };
            }
        }

        sealed class SolverState
        {
            public List<BottleSim> Bottles = new();
            public int Collected;
            public int[] PocketColors = new int[PocketCount];
            public bool[] PocketLocked = new bool[PocketCount];
            public List<int> PocketRefill = new();

            public SolverState Clone()
            {
                var clone = new SolverState
                {
                    Collected = Collected,
                    PocketColors = (int[])PocketColors.Clone(),
                    PocketLocked = (bool[])PocketLocked.Clone(),
                    PocketRefill = new List<int>(PocketRefill)
                };
                foreach (var b in Bottles)
                    clone.Bottles.Add(b.CloneShallow());
                return clone;
            }

            public bool CanPourFrom(int i)
            {
                var b = Bottles[i];
                return b.InPlay && !b.IsLocked && b.Stack.Count > 0 && !IsCollect(i);
            }

            public bool CanPourTo(int i)
            {
                var b = Bottles[i];
                return b.InPlay && !b.IsLocked && b.Stack.Count < MaxCapacity;
            }

            public bool CanPour(int from, int to) =>
                CanPourFrom(from) && CanPourTo(to) &&
                (Bottles[to].Stack.Count == 0 || Bottles[from].Stack[^1] == Bottles[to].Stack[^1]);

            public int TopPourRun(int from)
            {
                var stack = Bottles[from].Stack;
                if (stack.Count == 0) return 0;
                var color = stack[^1];
                var run = 0;
                for (var i = stack.Count - 1; i >= 0; i--)
                {
                    if (IsHiddenLayer(from, i)) break;
                    if (stack[i] != color) break;
                    run++;
                }

                return run;
            }

            bool IsHiddenLayer(int bottleIndex, int layerIndex)
            {
                var b = Bottles[bottleIndex];
                return b.Stack.Count > 1 && (b.WhMask & (1 << layerIndex)) != 0;
            }

            public bool IsCollect(int i)
            {
                var b = Bottles[i];
                if (!b.InPlay || b.Stack.Count != MaxCapacity || b.WhMask != 0) return false;
                var c = b.Stack[0];
                for (var l = 1; l < b.Stack.Count; l++)
                {
                    if (b.Stack[l] != c) return false;
                }

                return true;
            }

            public void ApplyPour(int from, int to, int move)
            {
                for (var m = 0; m < move; m++)
                {
                    var color = Bottles[from].Stack[^1];
                    Bottles[from].Stack.RemoveAt(Bottles[from].Stack.Count - 1);
                    Bottles[to].Stack.Add(color);
                }
            }

            public void ApplyAutoPackAndUnlock()
            {
                var guard = 0;
                while (guard++ < 64)
                {
                    if (TryPackOne())
                        continue;
                    if (!TryUnlockPocketForPendingCollect())
                        break;
                }
            }

            bool HasOpenPocketFor(int color)
            {
                for (var p = 0; p < PocketCount; p++)
                {
                    if (PocketLocked[p]) continue;
                    if (PocketColors[p] == color) return true;
                }

                return false;
            }

            bool TryUnlockPocketForPendingCollect()
            {
                for (var i = 0; i < Bottles.Count; i++)
                {
                    if (!IsCollect(i) || Bottles[i].IsLocked) continue;
                    var color = Bottles[i].Stack[^1];
                    if (HasOpenPocketFor(color)) continue;
                    return UnlockNextPocket();
                }

                return false;
            }

            bool UnlockNextPocket()
            {
                for (var p = 0; p < PocketCount; p++)
                {
                    if (!PocketLocked[p]) continue;
                    PocketLocked[p] = false;
                    PocketColors[p] = PocketRefill.Count > 0 ? PocketRefill[0] : 0;
                    if (PocketRefill.Count > 0)
                        PocketRefill.RemoveAt(0);
                    return true;
                }

                return false;
            }

            bool TryPackOne()
            {
                for (var p = 0; p < PocketCount; p++)
                {
                    if (PocketLocked[p]) continue;
                    var need = PocketColors[p];
                    if (need <= 0) continue;

                    for (var i = 0; i < Bottles.Count; i++)
                    {
                        if (!IsCollect(i) || IsLocked(i)) continue;
                        if (Bottles[i].Stack[^1] != need) continue;
                        PackBottle(i, p);
                        return true;
                    }
                }

                return false;
            }

            bool IsLocked(int i) => Bottles[i].IsLocked;

            void PackBottle(int bottleIndex, int pocketIndex)
            {
                var color = Bottles[bottleIndex].Stack[^1];
                Bottles[bottleIndex].Packed = true;
                Bottles[bottleIndex].Stack.Clear();
                Collected++;
                ApplyLockUnlockOnPack(color);
                PocketColors[pocketIndex] = PocketRefill.Count > 0 ? PocketRefill[0] : 0;
                if (PocketRefill.Count > 0)
                    PocketRefill.RemoveAt(0);
            }

            void ApplyLockUnlockOnPack(int packedColor)
            {
                foreach (var b in Bottles)
                {
                    if (!b.IsLockBottle || b.LockRemaining <= 0) continue;
                    if (b.LockColor != 0 && b.LockColor != packedColor) continue;
                    b.LockRemaining--;
                }
            }

            public string SerializeKey()
            {
                var sb = new StringBuilder(256);
                sb.Append('C').Append(Collected).Append(';');
                for (var p = 0; p < PocketCount; p++)
                    sb.Append(PocketColors[p]).Append(PocketLocked[p] ? 'L' : 'U').Append(',');
                sb.Append('|');
                foreach (var r in PocketRefill)
                    sb.Append(r).Append(',');
                sb.Append('|');
                foreach (var b in Bottles)
                {
                    sb.Append(b.Packed ? 'P' : 'A');
                    sb.Append(b.LockRemaining).Append(':');
                    foreach (var c in b.Stack)
                        sb.Append(c).Append(',');
                    sb.Append(';');
                }

                return sb.ToString();
            }
        }
    }
}
